middleware/internal/internal.go

```diff
@@ -0,0 +1,88 @@
+// The package internal provides a simple middleware that (a) prevents access
+// to internal locations and (b) allows to return files from internal location
+// by setting a special header, e.g. in a proxy response.
+package internal
+
+import (
+	"net/http"
+
+	"github.com/mholt/caddy/middleware"
+)
+
+// Internal middleware protects internal locations from external requests -
+// but allows access from the inside by using a special HTTP header.
+type Internal struct {
+	Next  middleware.Handler
+	Paths []string
+}
+
+const redirectHeader string = "X-Accel-Redirect"
+
+func isInternalRedirect(w http.ResponseWriter) bool {
+	return w.Header().Get(redirectHeader) != ""
+}
+
+// ServeHTTP implements the middlware.Handler interface.
+func (i Internal) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
+
+	// Internal location requested? -> Not found.
+	for _, prefix := range i.Paths {
+		if middleware.Path(r.URL.Path).Matches(prefix) {
+			return http.StatusNotFound, nil
+		}
+	}
+
+	// Use internal response writer to ignore responses that will be
+	// redirected to internal locations
+	iw := internalResponseWriter{ResponseWriter: w}
+	status, err := i.Next.ServeHTTP(iw, r)
+
+	if isInternalRedirect(iw) && status < 400 {
+		// Redirect - adapt request URL path and send it again
+		// "down the chain"
+		r.URL.Path = iw.Header().Get(redirectHeader)
+		iw.ClearHeader()
```

Comment:
Nginx with the following configuration:

```
events {
  worker_connections 768;
}

http {
  server {
    listen 8000;

    location /internal/ {
      internal;
      root /vagrant/;
    }

    location / {
      proxy_pass http://localhost:9000/;
    }
  }
}
```

Result: (1) the headers returned by the backend are cleared, (2) the internal redirection is done regardless of the HTTP status code returned by the backend and (3) Nginx supports multiple internal redirections.

I will adapt the code and also implement (2) and (3).